### ISSUE TITLE
Terraform 0.12.5 cannot output module values

### ISSUE THREAD

**Issue by jsirianni** [hashicorp/terraform#22126]
### Terraform Version

```
v0.12.5
```

### Terraform Configuration Files

output.tf in my module
```hcl
output "image" {
  value = var.image
}
```

main.tf in my root module
```hcl
module "dns_override" {
  source = "./mymodule"
}

```

### Expected Behavior

In the past, `terraform output -module=mymodule` would give me all outputs in a given module. 

### Actual Behavior

The documentations suggests that new syntax would work. `terraform output module.mymodule.image`, however, I get this:

```
The state file either has no outputs defined, or all the defined
outputs are empty. Please define an output in your configuration
with the `output` keyword and run `terraform refresh` for it to
become available. If you are using interpolation, please verify
the interpolated value is not empty. You can use the
`terraform console` command to assist.
```

The documentation also suggest that only outputs define din the root module, are available with the cli. Any outputs in a module are available for the root module to use / reference. 

### Steps to Reproduce
1) use `terraform output` to retrieve a single output (or all outputs) of a module used by your root module. 

### Additional Context

My tooling relies on `terraform output` to retrieve info about specific modules in use. If I cannot get outputs from the modules, I will have to significantly rework my test suite and readme instructions on my repositories. 

I suppose I could define outputs in my root module, but it would be very annoying as my root modules only purpose is to import my custom module a handful of times (I use the same module about 13 times in main.tf)

**Comment by teamterraform** [comment]
Hi @jsirianni,

Terraform 0.12 and later intentionally track only root module outputs in the state. To expose module outputs for external consumption, you must export them from the root module using an `output` block, which as of 0.12 can now be done for a single module all in one output, like this:

```hcl
output "example_module" {
  value = module.example_module
}
```

The outputs of child modules only exist temporarily in memory during a Terraform run, and are thus not available for direct consumption in `terraform output`. This is symmetrical with the idea that the CLI is only able to set values for input variables in the root module, and so input variables to child modules must always be explicitly passed through in a `module` block. Terraform internally now treats input variables, local values, and output values all the same way, and so only the ones defined on the root module are "special" in the sense that they are exposed out for the CLI to control/access. Only the root module defines the public interface for the configuration; everything else is an implementation detail.

Because of this, you will indeed unfortunately need to adjust your testing strategy to only test the public interface of the configuration, not its internals. Terraform no longer tracks the module outputs as part of a saved state snapshot -- they exist only in the internals of the configuration languge engine -- so the data you are trying to access isn't there to access.

---

The error message you got here was confusing, but I think that's because the correct way to do what you were doing here in 0.11 was to use the `-module` option, giving the following command line:

```
terraform output -module=mymodule image
```

If you do that in Terraform 0.12, Terraform should explain the same thing we described above:

```
Error: Unsupported option

The -module option is no longer supported since Terraform 0.12, because now
only root outputs are persisted in the state.
```

I think you got the less helpful message because Terraform was preparing to look for an output in the root module named `module.mymodule.image`. There is a missing check in the `terraform output` implementation to verify that the given output name is syntactically valid; it should have reported that `module.mymodule.image` is not a suitable output name, because output names cannot contain dots. We've created #22127 to represent that error message bug.
